sdk/client.go

```diff
@@ -0,0 +1,141 @@
+package sdk
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+
+	healthcheck "github.com/ONSdigital/dp-api-clients-go/v2/health"
+	health "github.com/ONSdigital/dp-healthcheck/healthcheck"
+	apiError "github.com/ONSdigital/dp-topic-api/sdk/errors"
+)
+
+const (
+	service = "dp-topic-api"
+)
+
+type Client struct {
+	hcCli *healthcheck.Client
+}
+
+// New creates a new instance of Client with a given topic api url
+func New(topicAPIURL string) *Client {
+	return &Client{
+		hcCli: healthcheck.NewClient(service, topicAPIURL),
+	}
+}
+
+// NewWithHealthClient creates a new instance of topic API Client,
+// reusing the URL and Clienter from the provided healthcheck client
+func NewWithHealthClient(hcCli *healthcheck.Client) (*Client, error) {
+	if hcCli == nil {
+		return nil, errors.New("health client is nil")
+	}
+	return &Client{
+		hcCli: healthcheck.NewClientWithClienter(service, hcCli.URL, hcCli.Client),
+	}, nil
+}
+
+// URL returns the URL used by this client
+func (cli *Client) URL() string {
+	return cli.hcCli.URL
+}
+
+// Health returns the underlying Healthcheck Client for this topic API client
+func (cli *Client) Health() *healthcheck.Client {
+	return cli.hcCli
+}
+
+// Checker calls topic api health endpoint and returns a check object to the caller
+func (cli *Client) Checker(ctx context.Context, check *health.CheckState) error {
+	return cli.hcCli.Checker(ctx, check)
+}
+
+// callTopicAPI calls the Topic API endpoint given by path for the provided REST method, request headers, and body payload.
+// It returns the response body and any error that occurred.
+func (cli *Client) callTopicAPI(ctx context.Context, path, method string, headers Headers, payload []byte) ([]byte, error) {
+	URL, err := url.Parse(path)
+	if err != nil {
+		return nil, apiError.StatusError{
+			Err:  fmt.Errorf("failed to parse path: \"%v\" error is: %v", path, err),
+			Code: http.StatusInternalServerError,
+		}
+	}
+
+	path = URL.String()
+
+	var req *http.Request
+
+	if payload != nil {
+		req, err = http.NewRequest(method, path, bytes.NewReader(payload))
```

Comment:
This err needs to be handled (I know this is the same in the Search Reindex though).

Reply:
The error is handled on line 81

Reply:
Yes sorry I meant that it needs to be handled before line 75, which uses the value of req, otherwise it could be that req is null (if an error has occurred) and then there would be a null pointer exception here: 		req.Header.Add("Content-type", "application/json")

Reply:
Amazing spot. Done